### PR TITLE
Pubsub executor thread pool size fix

### PR DESCRIPTION
**Pull request by dashevchenko**
## Pull Request description

Made changes to pubsub queue/rule node implementation to use only one manageable executor with a fixed thread pool size to avoid creating too many threads. 
To manage the size of thread pool env variables were added to yml file.

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [ ] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

**Review by YevhenBondarenko — APPROVED** [review]

File: common/queue/src/main/java/org/thingsboard/server/queue/pubsub/TbPubSubProducerTemplate.java [hunk 1]
```diff
@@ -120,7 +121,14 @@ private Publisher getOrCreatePublisher(String topic) {
             try {
                 admin.createTopicIfNotExists(topic);
                 ProjectTopicName topicName = ProjectTopicName.of(pubSubSettings.getProjectId(), topic);
-                Publisher publisher = Publisher.newBuilder(topicName).setCredentialsProvider(pubSubSettings.getCredentialsProvider()).build();
+                InstantiatingExecutorProvider executorProvider =
```
Comment (smatvienko-tb) [inline common/queue/src/main/java/org/thingsboard/server/queue/pubsub/TbPubSubProducerTemplate.java:124]: Please check the builder uses the executor provider

image

As far as I see,  the `InstantiatingExecutorProvider` will create a new thread pool all the time you create a producer.
I prefer to have only an instance of a thread pool.

image

InstantiatingExecutorProvider

image

I believe the `FixedExecutorProvider` is a proper choice. The `ScheduledExecutorService` you probably should create and control yourself. You can found many examples in the codebase.

image

**Review by smatvienko-tb — CHANGES_REQUESTED** [review]
Hi!
Good progress!
Please find my comment below to improve.

File: application/src/main/java/org/thingsboard/server/service/executors/PubSubRuleNodeExecutorProvider.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    /**
+    * Refers to com.google.cloud.pubsub.v1.Publisher default executor configuration
+    */
+    private static final int THREADS_PER_CPU = 5;
+    private ScheduledExecutorService executor;
+
+    @PostConstruct
+    public void init() {
+        executor = Executors.newScheduledThreadPool(THREADS_PER_CPU * Runtime.getRuntime().availableProcessors(), ThingsBoardThreadFactory.forName("pubsub-rule-nodes"));;
```
Comment (smatvienko-tb) [inline application/src/main/java/org/thingsboard/server/service/executors/PubSubRuleNodeExecutorProvider.java:22]: It should be PreDestroy with executor.shutdown

File: application/src/main/resources/thingsboard.yml [hunk 3]
```diff
@@ -1411,6 +1411,8 @@ queue:
     max_msg_size: "${TB_QUEUE_PUBSUB_MAX_MSG_SIZE:1048576}"
     # Number of messages per consumer
     max_messages: "${TB_QUEUE_PUBSUB_MAX_MESSAGES:1000}"
+    # Thread pool size for pubsub queue executor provider. If not set - default pubsub executor provider value will be used (5 * number of available processors)
```
Comment (ashvayka) [inline application/src/main/resources/thingsboard.yml:1414]: @dashevchenko , @smatvienko-tb ,  why 5xCPU count? Is this a recommended value?

File: common/queue/src/main/java/org/thingsboard/server/queue/pubsub/TbPubSubProducerTemplate.java [hunk 4]
```diff
@@ -120,7 +121,10 @@ private Publisher getOrCreatePublisher(String topic) {
             try {
                 admin.createTopicIfNotExists(topic);
                 ProjectTopicName topicName = ProjectTopicName.of(pubSubSettings.getProjectId(), topic);
-                Publisher publisher = Publisher.newBuilder(topicName).setCredentialsProvider(pubSubSettings.getCredentialsProvider()).build();
+                Publisher publisher = Publisher.newBuilder(topicName)
+                        .setCredentialsProvider(pubSubSettings.getCredentialsProvider())
+                        .setExecutorProvider(FixedExecutorProvider.create(pubSubSettings.getTbPubSubQueueExecutorProvider().getExecutor()))
```
Comment (ashvayka) [inline common/queue/src/main/java/org/thingsboard/server/queue/pubsub/TbPubSubProducerTemplate.java:126]: This change is not consistent with the way how we set the CredentialsProvider. PubSubSettings already has the `getCredentialsProvider()` method but does not have the `getExecutorProvider()`

File: common/queue/src/main/java/org/thingsboard/server/queue/pubsub/TbPubSubQueueExecutorProvider.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+import javax.annotation.PostConstruct;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+@ConditionalOnExpression("'${queue.type:null}'=='pubsub'")
+@Component
+public class TbPubSubQueueExecutorProvider implements ExecutorProvider {
+
+    @Value("${queue.pubsub.executor_thread_pool_size}")
```
[the remark this answers is not in the document]
Comment (ashvayka) [reply common/queue/src/main/java/org/thingsboard/server/queue/pubsub/TbPubSubQueueExecutorProvider.java:33]: Totally agree

File: application/src/main/java/org/thingsboard/server/service/executors/PubSubRuleNodeExecutorProvider.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    */
+    private static final int THREADS_PER_CPU = 5;
+    private ScheduledExecutorService executor;
+
+    @PostConstruct
+    public void init() {
+        if (threadPoolSize == null) {
+            threadPoolSize = THREADS_PER_CPU * Runtime.getRuntime().availableProcessors();
+        }
+        executor = Executors.newScheduledThreadPool(threadPoolSize, ThingsBoardThreadFactory.forName("pubsub-rule-nodes"));
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/executors/PubSubRuleNodeExecutorProvider.java:49]: Used not only for rule nodes but also for queues, right? That is why the name is not correct.

File: application/src/main/resources/thingsboard.yml (hunk 3 above)
Comment (dashevchenko) [reply application/src/main/resources/thingsboard.yml:1414]: this is default thread pool size value for default provider

File: application/src/main/java/org/thingsboard/server/actors/ActorSystemContext.java [hunk 7]
```diff
@@ -322,6 +323,11 @@ public ConcurrentMap<TenantId, DebugTbRateLimits> getDebugPerTenantLimits() {
     @Getter
     private NotificationExecutorService notificationExecutor;
 
+    @Lazy
```
[the remark this answers is not in the document]
Comment (smatvienko-tb) [reply application/src/main/java/org/thingsboard/server/actors/ActorSystemContext.java:326]: If Scheduled Pool starts with 0 threads, you can create that pool with no overhead, Please check the JavaDoc for more details

**Pull request merged**